File: tracer-python/app/tracer.py
"""
Execution tracer built on sys.monitoring (PEP 669, Python 3.12+).

Why sys.monitoring rather than sys.settrace -- the two reasons this tool can do
things Python Tutor cannot:

1. Speed. settrace calls back into Python for every line of every frame,
   including library internals, and costs 10-100x. sys.monitoring instruments
   bytecode directly and is roughly an order of magnitude cheaper.

2. DISABLE. A callback can return sys.monitoring.DISABLE to permanently retire
   *that code location* from monitoring. We use it to skip library code after
   its first hit, so importing pandas costs a handful of callbacks total rather
   than millions. That is what makes `import pandas` viable at all, and it has
   no equivalent under settrace.

The result: user code is traced line by line, library code runs at full native
speed, and library *objects* still show up in the heap snapshot.
"""

from __future__ import annotations

import builtins
import io
import sys
import time
import traceback
from typing import Any

from .limits import Limits
from .snapshot import Snapshotter

MON = sys.monitoring
TOOL_ID = MON.DEBUGGER_ID

# Filename we compile user code under. Everything else is "library".
USER_FILE = "<snippet>"


class StepBudgetExceeded(Exception):
    """Raised inside user code to unwind once the step cap is hit."""


class Timeout(Exception):
    """Raised inside user code when the wall clock runs out."""


class Tracer:
    def __init__(self, source: str, limits: Limits) -> None:
        self.source = source
        self.limits = limits
        self.snapshotter = Snapshotter(limits)

        self.steps: list[dict] = []
        self.stdout = io.StringIO()
        self.error: dict | None = None
        self.status = "ok"

        self._frame_ids: dict[int, str] = {}
        self._frame_counter = 0
        self._deadline = 0.0
        self._stopping = False

    # -- frame identity -----------------------------------------------------

    def _frame_id(self, frame: Any) -> str:
        """
        Stable id per *invocation*.

        Recursion means several live frames share a function name, so the name
        alone is not enough -- without a per-invocation counter the UI cannot
        tell fact(4) from fact(3) and animates them as one box.
        """
        key = id(frame)
        existing = self._frame_ids.get(key)
        if existing is None:
            self._frame_counter += 1
            name = frame.f_code.co_name
            label = "global" if name == "<module>" else name
            existing = f"{label}#{self._frame_counter}"
            self._frame_ids[key] = existing
        return existing

    def _user_frames(self) -> list[Any]:
        """Live user frames, outermost first. Library frames are excluded."""
        out = []
        frame = sys._getframe()
        while frame is not None:
            if frame.f_code.co_filename == USER_FILE:
                out.append(frame)
            frame = frame.f_back
        out.reverse()
        return out

    # -- step recording -----------------------------------------------------

    def _record(self, event: str, line: int, returned: Any = None,
                has_return: bool = False) -> None:
        if self._stopping:
            return

        if len(self.steps) >= self.limits.max_steps:
            self.status = "truncated"
            self._stopping = True
            raise StepBudgetExceeded()

        if time.monotonic() > self._deadline:
            self.status = "timeout"
            self._stopping = True
            raise Timeout()

        frames_raw = self._user_frames()
        if not frames_raw:
            return

        heap: dict[str, dict] = {}
        frames = [
            self.snapshotter.encode_frame(
                self._frame_id(f), f, heap, is_global=(f.f_code.co_name == "<module>")
            )
            for f in frames_raw
        ]

        step = {
            "i": len(self.steps),
            "line": line,
            "event": event,
            "stdout_len": self.stdout.tell(),
            "frames": frames,
            "heap": heap,
        }
        if has_return:
            step["returned"] = self.snapshotter.encode_value(returned, heap)
        self.steps.append(step)

    # -- monitoring callbacks ----------------------------------------------
    #
    # Each returns MON.DISABLE for non-user code. That is the load-bearing
    # optimisation: the location is retired from monitoring permanently, so
    # library code costs one callback ever rather than one per execution.

    def _on_line(self, code, line_number: int):
        if code.co_filename != USER_FILE:
            return MON.DISABLE
        self._record("line", line_number)
        return None

    def _on_start(self, code, _offset: int):
        if code.co_filename != USER_FILE:
            return MON.DISABLE
        # The module frame's "call" is the program starting, which is noise.
        if code.co_name == "<module>":
            return None
        self._record("call", code.co_firstlineno)
        return None

    def _on_return(self, code, _offset: int, retval: Any):
        if code.co_filename != USER_FILE:
            return MON.DISABLE

        frames = self._user_frames()
        line = frames[-1].f_lineno if frames else code.co_firstlineno

        if code.co_name == "<module>":
            # LINE events fire BEFORE their line runs, so the effect of the very
            # last statement is never captured by one. The module's return is
            # the only moment where the program has finished but its frame is
            # still alive, so this is where the final state gets recorded --
            # without it, `b.next = a` on the last line appears to do nothing.
            self._record("line", line)
            return None

        self._record("return", line, returned=retval, has_return=True)
        return None

    def _on_raise(self, code, _offset: int, exc: BaseException):
        # RAISE does NOT support DISABLE -- PEP 669 allows it only for local
        # events (LINE, PY_START, PY_RETURN, CALL, ...). Returning DISABLE here
        # raises ValueError from inside the callback, which then masquerades as
        # the user's own exception and corrupts every error, timeout and
        # truncation result. Library raises are rare, so plain filtering costs
        # nothing measurable.
        if code.co_filename != USER_FILE:
            return None
        # Our own control-flow signals must not be recorded as user exceptions.
        if isinstance(exc, (StepBudgetExceeded, Timeout)):
            return None
        frames = self._user_frames()
        line = frames[-1].f_lineno if frames else code.co_firstlineno
        self._record("exception", line)
        return None

    # -- run ----------------------------------------------------------------

    def run(self, stdin: str = "") -> dict:
        compiled = self._compile()
        if compiled is None:
            return self.to_document()

        globals_: dict[str, Any] = {
            "__name__": "__main__",
            "__file__": USER_FILE,
            "__builtins__": builtins,
        }

        self._deadline = time.monotonic() + self.limits.timeout_ms / 1000
        started = time.monotonic()

        real_stdout, real_stderr, real_stdin = sys.stdout, sys.stderr, sys.stdin
        sys.stdout = self.stdout
        sys.stderr = self.stdout
        sys.stdin = io.StringIO(stdin)

        try:
            self._install()
            try:
                exec(compiled, globals_)
            except StepBudgetExceeded:
                pass  # status already set to "truncated"
            except Timeout:
                pass  # status already set to "timeout"
            except BaseException as exc:  # noqa: BLE001 - user code raises anything
                self._capture_error(exc)
        finally:
            self._uninstall()
            self._capture_figures()
            sys.stdout, sys.stderr, sys.stdin = real_stdout, real_stderr, real_stdin

        self.duration_ms = (time.monotonic() - started) * 1000
        return self.to_document()

    def _capture_figures(self) -> None:
        """
        Attach any matplotlib plot the user built to the final step.

        Detected via sys.modules rather than an import, so a run that never
        touches matplotlib pays nothing. The sandbox image forces the Agg
        backend (MPLBACKEND=Agg), so figures render to a buffer with no display.
        Only the highest-numbered (most recent) figure is attached -- the schema
        carries one figure per step and this is the plot the user is looking at.
        """
        if "matplotlib.pyplot" not in sys.modules or not self.steps:
            return
        try:
            import base64
            plt = sys.modules["matplotlib.pyplot"]
            nums = plt.get_fignums()
            if not nums:
                return
            fig = plt.figure(nums[-1])
            buf = io.BytesIO()
            fig.savefig(buf, format="png", dpi=90, bbox_inches="tight")
            self.steps[-1]["figure"] = base64.b64encode(buf.getvalue()).decode("ascii")
            plt.close("all")
        except Exception:  # noqa: BLE001 - a plotting failure must not lose the trace
            pass

    def _compile(self):
        try:
            return compile(self.source, USER_FILE, "exec")
        except SyntaxError as exc:
            self.status = "compile_error"
            self.error = {
                "type": "SyntaxError",
                "message": exc.msg or "invalid syntax",
                "line": exc.lineno or 1,
            }
            return None

    def _capture_error(self, exc: BaseException) -> None:
        self.status = "error"
        frames = [
            {"name": f.name, "line": f.lineno}
            for f in traceback.extract_tb(exc.__traceback__)
            if f.filename == USER_FILE
        ]
        self.error = {
            "type": type(exc).__name__,
            "message": str(exc)[: self.limits.max_string],
            "line": frames[-1]["line"] if frames else None,
            "traceback": frames,
        }

    def _install(self) -> None:
        MON.use_tool_id(TOOL_ID, "visualizer")
        events = MON.events
        MON.register_callback(TOOL_ID, events.LINE, self._on_line)
        MON.register_callback(TOOL_ID, events.PY_START, self._on_start)
        MON.register_callback(TOOL_ID, events.PY_RETURN, self._on_return)
        MON.register_callback(TOOL_ID, events.RAISE, self._on_raise)
        MON.set_events(
            TOOL_ID,
            events.LINE | events.PY_START | events.PY_RETURN | events.RAISE,
        )

    def _uninstall(self) -> None:
        try:
            MON.set_events(TOOL_ID, 0)
            for event in (MON.events.LINE, MON.events.PY_START,
                          MON.events.PY_RETURN, MON.events.RAISE):
                MON.register_callback(TOOL_ID, event, None)
            MON.free_tool_id(TOOL_ID)
        except Exception:  # noqa: BLE001
            # Never let teardown mask the user's actual result.
            pass

    # -- output -------------------------------------------------------------

    def to_document(self) -> dict:
        doc = {
            "version": 1,
            "language": "python",
            "status": self.status,
            "source": self.source,
            "steps": self.steps,
            "stdout": self.stdout.getvalue(),
            "limits": self.limits.as_dict(),
            "meta": {
                "step_count": len(self.steps),
                "duration_ms": round(getattr(self, "duration_ms", 0.0), 2),
                "runtime_version": (
                    f"{sys.version_info.major}.{sys.version_info.minor}"
                    f".{sys.version_info.micro}"
                ),
                "packages": available_packages(),
                # How Python spells these. The replayer must never branch on
                # `language`, and "None" vs "null" vs "nil" is the one place
                # where rendering a value depends on where it came from.
                "literals": {"null": "None", "true": "True", "false": "False"},
            },
        }
        if self.error:
            doc["error"] = self.error
        return doc


import importlib.util as _importlib_util

_DISPLAY_NAMES = {"sklearn": "scikit-learn"}
_KNOWN_PACKAGES = ("numpy", "pandas", "matplotlib", "scipy", "sklearn")
_available_cache: list[str] | None = None


def available_packages() -> list[str]:
    """
    Third-party packages present in the image, for display in the UI.

    Uses find_spec, NOT __import__: this runs inside to_document() on every
    trace, in a freshly spawned child where nothing is cached. Actually
    importing the scientific stack here added ~1-2s to every run, including
    `x = 1`. find_spec only checks that the module is importable.
    """
    global _available_cache
    if _available_cache is None:
        found = []
        for name in _KNOWN_PACKAGES:
            try:
                if _importlib_util.find_spec(name) is not None:
                    found.append(_DISPLAY_NAMES.get(name, name))
            except (ImportError, ValueError):
                pass
        _available_cache = found
    return _available_cache


def trace(source: str, stdin: str = "", limits: Limits | None = None) -> dict:
    return Tracer(source, limits or Limits()).run(stdin)

File: tracer-node/src/tracer.js
/**
 * JavaScript / TypeScript tracer.
 *
 * Same shape as tracer-java: the snippet runs in a *separate* process under
 * debug control and this one drives it, so user code cannot take the service
 * down and a runaway loop is a `kill` away. There, the debugger is JDI; here it
 * is the V8 inspector over CDP.
 *
 * The hard problem is the same one every step-through tool has -- staying out
 * of the standard library. `Debugger.setBlackboxPatterns` does not help: Node
 * compiles its internals during bootstrap, before the inspector attaches, so
 * they are never re-evaluated against the patterns. One `console.log` will drag
 * a naive stepper through a dozen internal frames.
 *
 * So the loop below never descends. It steps into, and the moment it finds
 * itself outside the user's file it steps straight back out, which costs one
 * round trip per library call instead of one per library *line*. Callbacks
 * handed to library code (`xs.map(x => ...)`) would be skipped by that rule, so
 * every line of the snippet also carries a breakpoint: the step loop stays out
 * of the library, and the breakpoints catch user code the library calls back
 * into.
 */

import { spawn } from "node:child_process";
import { writeFile } from "node:fs/promises";
import path from "node:path";

import { Cdp } from "./cdp.js";
import { bootstrapSource } from "./bootstrap.js";
import { PACKAGES_DIR, availablePackages } from "./packages.js";
import { transpile } from "./typescript.js";

/**
 * How the replayer should print `null`, `true` and `false` for this language.
 * Sent with the document so the UI never has to branch on which language
 * produced it -- "None" versus "null" is a fact about the source language, and
 * the tracer is the only party that knows it.
 */
const LITERALS = { null: "null", true: "true", false: "false" };

/** Scope kinds that hold the user's own bindings. `global` and `closure` are
 *  deliberately absent: one dumps the entire runtime, the other repeats a
 *  parent frame's variables inside every child. */
const USER_SCOPES = new Set(["local", "block", "catch", "module", "with"]);

const CONNECT_TIMEOUT_MS = 5000;

export async function trace({ source, language = "javascript", stdin = "", limits, dir }) {
  const started = Date.now();
  const program = prepare(source, language);
  if (program.error) {
    return document({ source, language, status: "compile_error", limits, started, error: program.error });
  }

  const entry = path.join(dir, program.filename);
  await writeFile(entry, program.code, "utf8");

  const child = spawn(process.execPath, ["--inspect-brk=127.0.0.1:0", entry], {
    cwd: dir,
    stdio: ["pipe", "pipe", "pipe"],
    // A curated environment, not the tracer's: traced code has no business
    // reading whatever happens to be exported into this process.
    env: {
      NODE_PATH: PACKAGES_DIR,
      NODE_NO_WARNINGS: "1",
      NO_COLOR: "1",
      PATH: "/usr/bin:/bin",
    },
  });

  const collected = { stdout: "", stderr: "" };
  child.stdout.on("data", (chunk) => { collected.stdout += chunk; });
  child.stderr.on("data", (chunk) => { collected.stderr += chunk; });
  child.on("error", () => { /* spawn failure surfaces as a missing ws url */ });
  if (stdin) child.stdin.write(stdin);
  child.stdin.end();

  const exited = new Promise((resolve) => child.on("exit", resolve));
  // "exit" can arrive with the last of stderr still in flight, and stderr is
  // where a parse error is reported. "close" is the event that means drained.
  const drained = new Promise((resolve) => child.on("close", resolve));
  let finished = false;
  exited.then(() => { finished = true; });

  const kill = () => { if (!child.killed) child.kill("SIGKILL"); };
  const deadline = started + limits.timeout_ms;
  const guard = setTimeout(kill, limits.timeout_ms + 2000);

  try {
    const wsUrl = await inspectorUrl(child, exited);
    if (!wsUrl) {
      // The process died before the debugger came up: almost always a syntax
      // error, which Node reports on stderr and nowhere else.
      return document({
        source, language, status: "compile_error", limits, started,
        stdout: clean(collected.stdout, dir),
        error: syntaxError(collected.stderr, dir, program.lineMap),
      });
    }

    const cdp = await Cdp.connect(wsUrl, CONNECT_TIMEOUT_MS);
    const run = await stepThrough({ cdp, child, exited, program, limits, deadline });
    cdp.close();
    // A run with no steps may still have something to say. Node holds a failing
    // module's report back until the inspector lets go ("Waiting for the
    // debugger to disconnect"), so killing here would destroy the only account
    // of why nothing ran. A run that did produce steps is killed at once, as
    // before -- that is what stops a runaway loop.
    if (run.steps.length === 0) {
      await Promise.race([drained, new Promise((r) => setTimeout(r, 1500))]);
    }
    kill();
    await exited;

    // A snippet that never parsed leaves no steps behind, whether the debugger
    // broke inside the loader (CommonJS) or never stopped at all (ESM, which
    // fails while linking, before any frame exists). Either way Node's own
    // report on stderr is the only account of what went wrong, and calling that
    // a successful run of nothing would be the worst answer available.
    const unparsed = run.steps.length === 0
      ? parseFailure(collected.stderr, dir, program.lineMap)
      : null;

    return document({
      source, language, limits, started,
      status: unparsed ? "compile_error" : run.status,
      steps: run.steps,
      stdout: clean(collected.stdout, dir),
      error: unparsed ?? run.error,
    });
  } finally {
    clearTimeout(guard);
    if (!finished) kill();
  }
}

/** Turn the submitted source into something Node can execute. */
function prepare(source, language) {
  if (language === "typescript") {
    const result = transpile(source);
    if (result.error) return result;
    return { filename: "main.js", code: result.code, lineMap: result.lineMap };
  }
  // ESM and CommonJS cannot be told apart by the file extension alone here, so
  // the syntax decides: `import`/`export` at the top level means a module.
  const esm = /^\s*(import|export)[\s{*]/m.test(source);
  return { filename: esm ? "main.mjs" : "main.js", code: source, lineMap: null };
}

/** Node announces the inspector on stderr; that line is the handshake. */
function inspectorUrl(child, exited) {
  return new Promise((resolve) => {
    let buffer = "";
    const onData = (chunk) => {
      buffer += chunk;
      const match = buffer.match(/ws:\/\/\S+/);
      if (match) {
        child.stderr.off("data", onData);
        resolve(match[0]);
      }
    };
    child.stderr.on("data", onData);
    exited.then(() => resolve(null));
    setTimeout(() => resolve(null), CONNECT_TIMEOUT_MS);
  });
}

async function stepThrough({ cdp, child, exited, program, limits, deadline }) {
  const pauses = [];
  let waiting = null;
  let waitTimer = null;
  let ended = false;

  const deliver = (params) => {
    if (waiting) {
      clearTimeout(waitTimer);
      const resolve = waiting;
      waiting = null;
      resolve(params);
    } else if (params) {
      pauses.push(params);
    }
  };
  const finish = () => { ended = true; deliver(null); };

  cdp.on("Debugger.paused", deliver);
  cdp.on("__closed", finish);
  exited.then(finish);

  // `process.exit()` with an inspector attached does not exit: Node holds the
  // process open until the debugger disconnects, announcing it on stderr. That
  // line is the only signal that the program is over, and without it the run
  // sits here until the timeout and reports a hang that never happened.
  child.stderr.on("data", (chunk) => {
    if (String(chunk).includes("Waiting for the debugger to disconnect")) finish();
  });

  const nextPause = () => {
    if (pauses.length) return Promise.resolve(pauses.shift());
    if (cdp.closed || ended) return Promise.resolve(null);
    return new Promise((resolve) => {
      waiting = resolve;
      // Nothing may block past the deadline: a debuggee can wedge in a native
      // call where no pause will ever arrive.
      waitTimer = setTimeout(() => deliver(null), Math.max(0, deadline - Date.now()) + 500);
    });
  };

  const scripts = new Map();
  cdp.on("Debugger.scriptParsed", (p) => scripts.set(p.scriptId, p.url));

  await cdp.send("Runtime.enable");
  await cdp.send("Debugger.enable");
  await cdp.send("Debugger.setPauseOnExceptions", { state: "uncaught" });
  await cdp.send("Runtime.runIfWaitingForDebugger");

  const first = await nextPause();
  if (!first || !first.callFrames?.length) {
    return { status: "ok", steps: [] };
  }

  const scriptId = first.callFrames[0].location.scriptId;
  const url = scripts.get(scriptId) ?? "";

  // --inspect-brk normally stops on the snippet's first line. It stops somewhere
  // else in exactly one case: the snippet did not parse, so it was never entered
  // and what breaks instead is the SyntaxError being thrown inside Node's module
  // loader. Those are not the user's frames, and adopting their scriptId here
  // would report the loader's line numbers and internal function names as if
  // they were the snippet's. Hand it back to the stderr parser, which reads the
  // line out of Node's own report.
  if (!url.endsWith(program.filename)) {
    await cdp.send("Debugger.resume");
    return { status: "compile_error", steps: [], error: null, unparsed: true };
  }

  await cdp.send("Runtime.evaluate", {
    expression: bootstrapSource(limits),
    returnByValue: true,
    silent: true,
  });

  await setLineBreakpoints(cdp, url, program.code);

  const steps = [];
  const stack = [];      // one entry per live user frame, outermost first
  let frameCounter = 0;
  let error = null;
  let status = "ok";
  let paused = first;
  // Null until the first step: entering the module is not a "call" the reader
  // made, and tracer-python opens its traces on a `line` too.
  let previousDepth = null;

  while (paused) {
    if (Date.now() > deadline) { status = "timeout"; break; }
    if (steps.length >= limits.max_steps) { status = "truncated"; break; }

    const index = paused.callFrames.findIndex((f) => f.location.scriptId === scriptId);
    if (index === -1) {
      // Past the end of the snippet, or deep in machinery with nothing of ours
      // below. Let it run.
      await cdp.send("Debugger.resume");
      break;
    }
    if (index > 0) {
      // Inside a library call. Leave immediately rather than stepping through it.
      await cdp.send("Debugger.stepOut");
      paused = await nextPause();
      continue;
    }

    const userFrames = paused.callFrames.filter((f) => f.location.scriptId === scriptId);
    const top = userFrames[0];

    /**
     * The module's own return.
     *
     * Steps are recorded BEFORE their line runs, so the effect of the very last
     * statement is never captured by one. This is the only moment where the
     * program has finished but its frame is still alive, which makes it the
     * place to record the final state -- without it, `b.next = a` on the last
     * line appears to do nothing. tracer-python records the same step for the
     * same reason.
     */
    const isModuleReturn = userFrames.length === 1 && top.returnValue !== undefined;

    let line = mapLine(top.location.lineNumber + 1, program.lineMap);
    if (line === null) {
      // Compiler output, not the user's code -- unless it is the final state,
      // which is worth attributing to the last line the reader actually saw.
      if (!isModuleReturn) {
        await cdp.send("Debugger.stepInto");
        paused = await nextPause();
        continue;
      }
      line = steps.at(-1)?.line ?? 1;
    }

    const isException = paused.reason === "exception" || paused.reason === "promiseRejection";
    const event = isException ? "exception"
      : isModuleReturn ? "line"
        : top.returnValue !== undefined ? "return"
          : (previousDepth !== null && userFrames.length > previousDepth) ? "call"
            : "line";
    previousDepth = userFrames.length;

    const step = await snapshot({ cdp, userFrames, program, index: steps.length, line, event });
    // A module "returns" its exports object, which means nothing to the reader.
    if (top.returnValue !== undefined && !isModuleReturn) {
      const returned = await encodeRemote(cdp, top.returnValue, step.heap);
      if (returned) step.returned = returned;
    }
    steps.push(step);

    if (isException) {
      error = await describeThrown(cdp, paused.data, userFrames, program);
      status = "error";
      await cdp.send("Debugger.resume");
      break;
    }

    await cdp.send("Debugger.stepInto");
    paused = await nextPause();
  }

  // A module that fails to parse throws before a single statement runs. That is
  // a compile error, not a program that crashed, and calling it one lets the UI
  // point at the offending line instead of showing an empty trace.
  if (status === "error" && error?.type === "SyntaxError" && steps.length <= 1) {
    return { status: "compile_error", steps: [], error };
  }

  return { status, steps, error };

  /** Frame identity: minted on entry, retired on return, so a recursive call
   *  gets its own id and the UI can animate frames independently. */
  function syncStack(userFrames) {
    const outermostFirst = [...userFrames].reverse();
    while (stack.length > outermostFirst.length) stack.pop();
    outermostFirst.forEach((frame, i) => {
      const name = frame.functionName || "";
      if (!stack[i] || stack[i].name !== name) {
        stack[i] = { id: `f${++frameCounter}`, name };
      }
    });
    return outermostFirst;
  }

  async function snapshot({ cdp, userFrames, program, index, line, event }) {
    const outermostFirst = syncStack(userFrames);
    const heap = {};
    const frames = [];
    let outLen = null;

    for (let i = 0; i < outermostFirst.length; i++) {
      const frame = outermostFirst[i];
      const isModule = i === 0;
      const scopes = (frame.scopeChain ?? []).filter(
        (s) => USER_SCOPES.has(s.type) && s.object?.objectId,
      );
      const captured = await captureScopes(cdp, scopes, isModule);
      Object.assign(heap, captured.heap);
      if (captured.out_len !== null && captured.out_len !== undefined) outLen = captured.out_len;

      frames.push({
        id: stack[i].id,
        name: isModule ? "global" : (frame.functionName || "(anonymous)"),
        line: mapLine(frame.location.lineNumber + 1, program.lineMap) ?? line,
        locals: captured.locals,
        order: captured.order,
        ...(isModule ? { is_global: true } : {}),
      });
    }

    return {
      i: index,
      line,
      event,
      ...(outLen === null ? {} : { stdout_len: outLen }),
      frames,
      heap,
    };
  }
}

/**
 * A breakpoint on every line of the snippet.
 *
 * This is the net for user code that library code calls back into: the step
 * loop refuses to walk through `Array.prototype.map`, so without these the
 * callback body would never be seen. V8 resolves each request to the nearest
 * real statement and silently reports none for blank lines and comments.
 */
async function setLineBreakpoints(cdp, url, code) {
  if (!url) return;
  const lines = code.split("\n").length;
  for (let line = 0; line < lines; line++) {
    await cdp.send("Debugger.setBreakpointByUrl", { url, lineNumber: line, columnNumber: 0 })
      .catch(() => null);
  }
}

/** One round trip per frame: the debuggee walks the whole reachable graph. */
async function captureScopes(cdp, scopes, isModule) {
  const empty = { locals: {}, order: [], heap: {}, out_len: null };
  if (scopes.length === 0) return empty;
  const [first, ...rest] = scopes;
  const { result } = await cdp.send("Runtime.callFunctionOn", {
    objectId: first.object.objectId,
    functionDeclaration:
      "function () { const a = [...arguments]; const skip = a.pop(); " +
      "return __viz.snapshotScopes([this, ...a], skip); }",
    arguments: [...rest.map((s) => ({ objectId: s.object.objectId })), { value: isModule }],
    returnByValue: true,
    silent: true,
  });
  return result?.value ?? empty;
}

/**
 * Encode a value the inspector handed us as a RemoteObject.
 *
 * Primitives arrive complete and are encoded here; anything with an objectId is
 * handed back to the debuggee's encoder so there is one implementation of the
 * heap shape rather than two that can drift.
 */
async function encodeRemote(cdp, remote, heap) {
  if (!remote) return null;
  switch (remote.type) {
    case "undefined": return { prim: "undefined", prim_type: "null" };
    case "boolean": return { prim: remote.value, prim_type: "bool" };
    case "string": return { prim: String(remote.value), prim_type: "str" };
    case "bigint": return { prim: remote.unserializableValue ?? String(remote.value), prim_type: "int" };
    case "symbol": return { prim: remote.description ?? "Symbol()", prim_type: "str" };
    case "number": {
      if (remote.unserializableValue) {
        return { prim: remote.unserializableValue, prim_type: "float" };
      }
      return Number.isInteger(remote.value)
        ? { prim: remote.value, prim_type: "int" }
        : { prim: remote.value, prim_type: "float" };
    }
    default: break;
  }
  if (remote.subtype === "null") return { prim: null, prim_type: "null" };
  if (!remote.objectId) return null;

  const { result } = await cdp.send("Runtime.callFunctionOn", {
    objectId: remote.objectId,
    functionDeclaration: "function () { return __viz.encodeValue(this); }",
    returnByValue: true,
    silent: true,
  });
  if (!result?.value) return null;
  Object.assign(heap, result.value.heap);
  return result.value.value;
}

/** Build the error block for an uncaught throw. */
async function describeThrown(cdp, thrown, userFrames, program) {
  const top = userFrames[0];
  const error = {
    type: thrown?.className ?? "Error",
    message: "Uncaught",
    line: mapLine(top.location.lineNumber + 1, program.lineMap) ?? undefined,
    traceback: userFrames.map((f) => ({
      name: f.functionName || "global",
      line: mapLine(f.location.lineNumber + 1, program.lineMap) ?? 0,
    })).reverse(),
  };

  if (thrown?.objectId) {
    const { result } = await cdp.send("Runtime.callFunctionOn", {
      objectId: thrown.objectId,
      functionDeclaration:
        "function () { return { name: this && this.name, message: this && this.message }; }",
      returnByValue: true,
      silent: true,
    });
    if (result?.value) {
      error.type = result.value.name || error.type;
      error.message = String(result.value.message ?? "").slice(0, 500) || error.message;
    }
  } else if (thrown) {
    // `throw "a string"` is legal and depressingly common.
    error.type = "Thrown";
    error.message = String(thrown.description ?? thrown.value ?? "").slice(0, 500);
  }
  return error;
}

/** Generated line -> source line. Unmapped generated lines return null. */
function mapLine(line, lineMap) {
  if (!lineMap) return line;
  return lineMap.get(line) ?? null;
}

/**
 * Pull a syntax error out of Node's stderr.
 *
 * The report starts with an absolute path, which would leak the server's
 * filesystem layout into the UI, so the run directory is scrubbed everywhere
 * before anything is returned.
 */
function syntaxError(stderr, dir, lineMap) {
  const found = parseFailure(stderr, dir, lineMap);
  if (found) return found;
  const text = clean(stderr, dir);
  return {
    type: "SyntaxError",
    message: text.trim().split("\n")[0] || "The snippet could not be parsed.",
  };
}

/**
 * Node's report for a snippet that never parsed, or null if stderr holds no
 * such thing.
 *
 * A snippet can fail to parse without the debugger ever stopping in it, so
 * "no steps" alone cannot be read as "the program did nothing" -- the parse
 * error may be the only evidence anything happened, and it lives here.
 */
function parseFailure(stderr, dir, lineMap) {
  const text = clean(stderr, dir);
  const match = text.match(/^(\w*(?:Error|Exception)): (.*)$/m);
  const located = text.match(/^\S*main\.m?js:(\d+)/m);
  if (!match || !located) return null;
  const rawLine = Number(located[1]);
  return {
    type: match[1],
    message: match[2],
    line: mapLine(rawLine, lineMap) ?? rawLine,
  };
}

/** Strip the run directory and the inspector's own chatter. */
function clean(text, dir) {
  return text
    .split(dir + path.sep).join("")
    .split(dir).join("")
    .replace(/^Debugger listening on ws:\/\/\S+\r?\n?/gm, "")
    .replace(/^For help, see: https:\/\/nodejs\.org\/en\/docs\/inspector\r?\n?/gm, "")
    .replace(/^Debugger attached\.\r?\n?/gm, "")
    .replace(/^Waiting for the debugger to disconnect\.\.\.\r?\n?/gm, "");
}

function document({ source, language, status, limits, started, steps = [], stdout = "", error }) {
  // stdout_len is counted inside the debuggee; if it was killed mid-write the
  // parent may hold less text than the counter promises. Clamp rather than let
  // the UI slice past the end.
  for (const step of steps) {
    if (step.stdout_len === undefined) step.stdout_len = stdout.length;
    step.stdout_len = Math.min(step.stdout_len, stdout.length);
  }
  return {
    version: 1,
    language,
    status,
    source,
    steps,
    stdout,
    ...(error ? { error } : {}),
    limits,
    meta: {
      duration_ms: Date.now() - started,
      step_count: steps.length,
      runtime_version: process.versions.node,
      packages: availablePackages(),
      literals: LITERALS,
    },
  };
}
